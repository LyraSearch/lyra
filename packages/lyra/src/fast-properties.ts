/*
 * Adapred from: https://github.com/sindresorhus/to-fast-properties
 *
 * Original license:
 *
 * MIT License
 *
 * Copyright (c) Petka Antonov
 * Copyright (c) Benjamin Gruenbaum
 * Copyright (c) John-David Dalton
 * Copyright (c) Sindre Sorhus
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let fastProto: any = null;

function FastObject<T extends object>(object?: T): T {
  if (fastProto !== null && typeof fastProto.property) {
    const result = fastProto;
    fastProto = FastObject.prototype = null;
    return result;
  }

  fastProto = FastObject.prototype = object == null ? Object.create(null) : object;

  // @ts-expect-error this is a hack to make the type checker happy (written by Copilot LOL)
  return new FastObject();
}

const inlineCacheCutoff = 10;

for (let index = 0; index <= inlineCacheCutoff; index++) {
  FastObject();
}

export function insertWithFastProperties<T extends object>(object: T, propName: keyof T, value: T[keyof T]): T {
  object[propName] = value;
  toFastProperties(object);
  return object;
}

export default function toFastProperties<T extends object>(object?: T): T {
  return FastObject(object);
}
