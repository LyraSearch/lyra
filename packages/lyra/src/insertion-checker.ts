const kInsertions = Symbol("lyra.insertions");

const warn =
  "process" in globalThis
    ? process.emitWarning
    : function emitWarning(message: string, options: { code: string }) {
        console.warn(`[WARNING] [${options.code}] ${message}`);
      };

export function trackInsertion(_lyra: unknown) {
  const lyra = _lyra as object & { [kInsertions]?: number };

  if (typeof lyra[kInsertions] !== "number") {
    queueMicrotask(() => {
      lyra[kInsertions] = undefined;
    });

    lyra[kInsertions] = 0;
  }

  if (lyra[kInsertions]! > 1000) {
    warn(
      "Lyra's insert operation is synchronous. Please avoid inserting a large number of document in a single operation in order not to block the main thread.",
      { code: "LYRA0001" },
    );

    lyra[kInsertions] = -1;
  } else if (lyra[kInsertions] >= 0) {
    lyra[kInsertions]++;
  }
}
