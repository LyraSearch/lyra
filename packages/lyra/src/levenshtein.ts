export function levenshtein(a: string, b: string): number {
  /* c8 ignore next 3 */
  if (!a.length) {
    return b.length;
  }

  /* c8 ignore next 3 */
  if (!b.length) {
    return a.length;
  }

  let tmp;

  if (a.length > b.length) {
    tmp = a;
    a = b;
    b = tmp;
  }

  const row = Array.from({ length: a.length + 1 }, (_, i) => i);
  let val = 0;

  for (let i = 1; i <= b.length; i++) {
    let prev = i;

    for (let j = 1; j <= a.length; j++) {
      if (b[i - 1] === a[j - 1]) {
        val = row[j - 1];
      } else {
        val = Math.min(row[j - 1] + 1, Math.min(prev + 1, row[j] + 1));
      }

      row[j - 1] = prev;
      prev = val;
    }
    row[a.length] = prev;
  }

  return row[a.length];
}
