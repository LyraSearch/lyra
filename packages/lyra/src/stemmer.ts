import {
  PorterStemmer,
  PorterStemmerNl,
  PorterStemmerFr,
  PorterStemmerEs,
  PorterStemmerIt,
  PorterStemmerPt,
  PorterStemmerSv,
  PorterStemmerRu,
  PorterStemmerNo,
  Stemmer,
} from "natural";

export type Language = typeof SUPPORTED_LANGUAGES[number];

export const SUPPORTED_LANGUAGES = [
  "dutch",
  "english",
  "french",
  "italian",
  "norwegian",
  "portugese",
  "russian",
  "spanish",
  "swedish",
] as const;

export function stemArray(input: string[], language: Language): string[] {
  let stemmer: Stemmer;

  switch (language) {
    case "dutch":
      stemmer = PorterStemmerNl;
      break;
    case "english":
      stemmer = PorterStemmer;
      break;
    case "french":
      stemmer = PorterStemmerFr;
      break;
    case "italian":
      stemmer = PorterStemmerIt;
      break;
    case "norwegian":
      stemmer = PorterStemmerNo;
      break;
    case "portugese":
      stemmer = PorterStemmerPt;
      break;
    case "russian":
      stemmer = PorterStemmerRu;
      break;
    case "spanish":
      stemmer = PorterStemmerEs;
      break;
    case "swedish":
      stemmer = PorterStemmerSv;
      break;
    default:
      stemmer = PorterStemmer;
      break;
  }

  return input.map(word => stemmer.stem(word));
}
