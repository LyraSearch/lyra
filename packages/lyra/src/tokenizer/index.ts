import { Language } from "./languages";

const splitRegex: Record<Language, RegExp> = {
  dutch: /[^a-z0-9_'-]+/gim,
  english: /[^a-z0-9_'-]+/gim,
  french: /[^a-z0-9äâàéèëêïîöôùüûœç-]+/gim,
  italian: /[^a-z0-9_'-]+/gim,
  norwegian: /[^a-z0-9_æøåÆØÅäÄöÖüÜ]+/gim,
  portugese: /[^a-zà-úÀ-Ú]/gim,
  russian: /[^a-zа-яА-ЯёЁ]+/gim,
  spanish: /[^a-zA-Zá-úÁ-ÚñÑüÜ]+/gim,
  swedish: /[^a-z0-9_åÅäÄöÖüÜ-]+/gim,
};

export function tokenize(input: string, language: Language = "english") {
  const splitRule = splitRegex[language];
  const tokens = input.toLowerCase().split(splitRule);
  return Array.from(new Set(trim(tokens)));
}

function trim(text: string[]): string[] {
  while (text[text.length - 1] === "") {
    text.pop();
  }
  while (text[0] === "") {
    text.shift();
  }
  return text;
}
