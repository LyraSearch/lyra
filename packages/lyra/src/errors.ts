import { SUPPORTED_LANGUAGES } from "./stemmer";

function formatJSON(input: object) {
  return JSON.stringify(input, null, 2);
}

export function INVALID_SCHEMA_TYPE(type: string): string {
  return `Invalid schema type. Expected string or object, but got ${type}`;
}

export function INVALID_DOC_SCHEMA(expected: object, found: object): string {
  return `Invalid document structure. \nClass has been initialized with the following schema: \n\n${formatJSON(
    expected,
  )}\n\nbut found the following doc:\n\n${formatJSON(found)}`;
}

export function INVALID_PROPERTY(name: string, expected: string[]): string {
  return `Invalid property name. Expected a wildcard string ("*") or array containing one of the following properties: ${expected.join(
    ", ",
  )}, but got: ${name}`;
}

export function CANT_DELETE_DOC_NOT_FOUND(id: string): string {
  return `Document with ID ${id} does not exist.`;
}

export function CANT_DELETE_DOCUMENT(docID: string, key: string, token: string): string {
  return `Unable to delete document "${docID}" from index "${key}" on word "${token}".`;
}

export function UNSUPPORTED_NESTED_PROPERTIES(): string {
  return `Nested properties are not supported in this Lyra version, but will be in the future.`;
}

export function DOC_ID_DOES_NOT_EXISTS(id: string): string {
  return `Document with ID ${id} does not exists`;
}

export function LANGUAGE_NOT_SUPPORTED(lang: string): string {
  return `Language "${lang}" is not supported.\nSupported languages are:\n - ${SUPPORTED_LANGUAGES.join("\n - ")}`;
}

export function GETTER_SETTER_WORKS_ON_EDGE_ONLY(method: string): string {
  return `${method} works on edge only. Use edge: true in Lyra constructor to enable it.`;
}
