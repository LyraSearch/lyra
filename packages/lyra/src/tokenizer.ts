import { WordTokenizer } from "natural";
import { Language, stemArray } from "./stemmer";

const tokenizer = new WordTokenizer();

export function tokenize(input: string, language: Language = "english", enableStemming = true): string[] {
  const tokens = tokenizer.tokenize(input);

  if (!enableStemming) {
    return Array.from(new Set(tokens));
  }

  return Array.from(new Set(stemArray(tokens, language)));
}
