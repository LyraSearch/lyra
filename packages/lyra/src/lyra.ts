import type { queueAsPromised } from "fastq";
import { nanoid } from "nanoid";
import * as fastq from "fastq";
import { Trie } from "./prefix-tree/trie";
import * as ERRORS from "./errors";
import { tokenize } from "./tokenizer";
import { formatNanoseconds, getNanosecondsTime } from "./utils";
import { Language, SUPPORTED_LANGUAGES } from "./stemmer";

export type PropertyType = "string" | "number" | "boolean";

export type PropertiesSchema = {
  [key: string]: PropertyType | PropertiesSchema;
};

export type LyraProperties = {
  schema: PropertiesSchema;
  defaultLanguage?: Language;
};

export type LyraDocs = Map<string, object>;

export type SearchParams = {
  term: string;
  properties?: "*" | string[];
  limit?: number;
  offset?: number;
  exact?: boolean;
};

type LyraIndex = Map<string, Trie>;

type QueueDocParams = {
  id: string;
  doc: object;
  language: Language;
};

type SearchResult = Promise<{
  count: number;
  hits: object[];
  elapsed: string;
}>;

export class Lyra {
  private defaultLanguage: Language = "english";
  private schema: PropertiesSchema;
  private docs: LyraDocs = new Map();
  private index: LyraIndex = new Map();
  private queue: queueAsPromised<QueueDocParams> = fastq.promise(
    this,
    this._insert,
    1
  );

  constructor(properties: LyraProperties) {
    const defaultLanguage =
      (properties?.defaultLanguage?.toLowerCase() as Language) ?? "english";

    if (!SUPPORTED_LANGUAGES.includes(defaultLanguage)) {
      throw ERRORS.LANGUAGE_NOT_SUPPORTED(defaultLanguage);
    }

    this.defaultLanguage = defaultLanguage;
    this.schema = properties.schema;
    this.buildIndex(properties.schema);
  }

  private buildIndex(schema: PropertiesSchema) {
    for (const prop in schema) {
      const propType = typeof prop;

      if (propType === "string") {
        this.index.set(prop, new Trie());
      } else if (propType === "object") {
        // @todo nested properties will be supported with the next versions
        //this.buildIndex(prop as unknown as PropertiesSchema);
        throw ERRORS.UNSUPPORTED_NESTED_PROPERTIES();
      } else {
        throw ERRORS.INVALID_SCHEMA_TYPE(propType);
      }
    }
  }

  async search(
    params: SearchParams,
    language: Language = this.defaultLanguage
  ): SearchResult {
    const tokens = tokenize(params.term, language).values();
    const indices = this.getIndices(params.properties);
    const { limit = 10, offset = 0, exact = false } = params;
    const results: object[] = new Array({ length: limit });
    let totalResults = 0;

    const timeStart = getNanosecondsTime();

    let i = 0;
    let j = 0;

    for (const token of tokens) {
      for (const index of indices) {
        const documentIDs = await this.getDocumentIDsFromSearch({
          ...params,
          index: index,
          term: token,
          exact: exact,
        });

        totalResults += documentIDs.size;

        if (i >= limit) {
          break;
        }

        if (documentIDs.size) {
          for (const id of documentIDs) {
            if (j < offset) {
              j++;
              continue;
            }

            if (i >= limit) {
              break;
            }

            const fullDoc = this.docs.get(id);
            results[i] = { id, ...fullDoc };
            i++;
          }
        }
      }
    }

    return {
      elapsed: formatNanoseconds(getNanosecondsTime() - timeStart),
      hits: results,
      count: totalResults,
    };
  }

  private getIndices(indices: SearchParams["properties"]): string[] {
    const knownIndices = [...this.index.keys()];

    if (!indices) {
      return knownIndices;
    }

    if (typeof indices === "string") {
      if (indices === "*") {
        return knownIndices;
      } else {
        throw ERRORS.INVALID_PROPERTY(indices, knownIndices);
      }
    }

    for (const index of indices as string[]) {
      if (!knownIndices.includes(index)) {
        throw ERRORS.INVALID_PROPERTY(index, knownIndices);
      }
    }

    return indices as string[];
  }

  async delete(docID: string): Promise<boolean> {
    if (!this.docs.has(docID)) {
      throw ERRORS.DOC_ID_DOES_NOT_EXISTS(docID);
    }

    const document = this.docs.get(docID)!;

    for (const key in document) {
      const idx = this.index.get(key)!;
      const tokens = tokenize((document as any)[key]);

      for (const token of tokens) {
        if (idx.removeDocByWord(token, docID)) {
          throw `Unable to remove document "${docID}" from index "${key}" on word "${token}".`;
        }
      }
    }

    this.docs.delete(docID);

    return true;
  }

  private async getDocumentIDsFromSearch(
    params: SearchParams & { index: string }
  ): Promise<Set<string>> {
    const idx = this.index.get(params.index);
    const searchResult = idx?.find({
      term: params.term,
      exact: params.exact,
    });
    const ids = new Set<string>();

    for (const key in searchResult) {
      for (const id of (searchResult as any)[key]) {
        ids.add(id);
      }
    }
    return ids;
  }

  public async insert(
    doc: object,
    language: Language = this.defaultLanguage
  ): Promise<{ id: string }> {
    const id = nanoid();

    if (!SUPPORTED_LANGUAGES.includes(language)) {
      throw ERRORS.LANGUAGE_NOT_SUPPORTED(language);
    }

    if (!(await this.checkInsertDocSchema(doc))) {
      throw ERRORS.INVALID_DOC_SCHEMA(this.schema, doc);
    }

    await this.queue.push({
      id,
      doc,
      language,
    });

    return { id };
  }

  private async _insert({ doc, id, language }: QueueDocParams): Promise<void> {
    const index = this.index;
    this.docs.set(id, doc);

    function recursiveTrieInsertion(doc: object) {
      for (const key in doc) {
        if (typeof (doc as any)[key] === "object") {
          throw ERRORS.UNSUPPORTED_NESTED_PROPERTIES();
          // @todo nested properties will be supported with the nexrt versions
          // recursiveTrieInsertion((doc as any)[key]);
        } else if (typeof (doc as any)[key] === "string") {
          const requestedTrie = index.get(key);
          const tokens = tokenize((doc as any)[key], language);

          for (const token of tokens) {
            requestedTrie?.insert(token, id);
          }
        }
      }
    }

    recursiveTrieInsertion(doc);
  }

  private async checkInsertDocSchema(
    doc: QueueDocParams["doc"]
  ): Promise<boolean> {
    function recursiveCheck(
      newDoc: QueueDocParams["doc"],
      schema: PropertiesSchema
    ): boolean {
      for (const key in newDoc) {
        if (!(key in schema)) {
          return false;
        }

        const propType = typeof (newDoc as any)[key];

        if (propType === "object") {
          recursiveCheck((newDoc as any)[key], schema);
        } else {
          if (typeof (newDoc as any)[key] !== schema[key]) {
            return false;
          }
        }
      }

      return true;
    }

    return recursiveCheck(doc, this.schema);
  }
}
