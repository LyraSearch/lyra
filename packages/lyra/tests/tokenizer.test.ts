import { stemArray } from "../src/stemmer";
import { tokenize } from "../src/tokenizer";

it("Should tokenize and stem correctly in english", () => {
  const I1 = "the quick brown fox jumps over the lazy dog";
  const I2 = "I baked some cakes";

  const O1 = tokenize(I1, "english");
  const O2 = tokenize(I2, "english");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it("Should tokenize and stem correctly in french", () => {
  const I1 = "voyons quel temps il fait dehors";
  const I2 = "j'ai fait des gâteaux";

  const O1 = tokenize(I1, "french");
  const O2 = tokenize(I2, "french");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it("Should tokenize and stem correctly in italian", () => {
  const I1 = "ho cucinato delle torte";
  const I2 = "dormire è una cosa difficile quando i test non passano";

  const O1 = tokenize(I1, "italian");
  const O2 = tokenize(I2, "italian");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it("Should tokenize and stem correctly in norwegian", () => {
  const I1 = "Jeg kokte noen kaker";
  const I2 = "å sove er en vanskelig ting når testene mislykkes";

  const O1 = tokenize(I1, "norwegian");
  const O2 = tokenize(I2, "norwegian");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it("Should tokenize and stem correctly in portugese", () => {
  const I1 = "Eu cozinhei alguns bolos";
  const I2 = "dormir é uma coisa difícil quando os testes falham";

  const O1 = tokenize(I1, "portugese");
  const O2 = tokenize(I2, "portugese");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it("Should tokenize and stem correctly in russian", () => {
  const I1 = "я приготовила пирожные";
  const I2 = "спать трудно, когда тесты не срабатывают";

  const O1 = tokenize(I1, "russian");
  const O2 = tokenize(I2, "russian");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it("Should tokenize and stem correctly in swedish", () => {
  const I1 = "Jag lagade några kakor";
  const I2 = "att sova är en svår sak när testerna misslyckas";

  const O1 = tokenize(I1, "swedish");
  const O2 = tokenize(I2, "swedish");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it.skip("Should tokenize and stem correctly in spanish", () => {
  const I1 = "cociné unos pasteles";
  const I2 = "dormir es algo dificil cuando las pruebas fallan";

  const O1 = tokenize(I1, "spanish");
  const O2 = tokenize(I2, "spanish");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it.skip("Should tokenize and stem correctly in dutch", () => {
  const I1 = "de kleine koeien";
  const I2 = "Ik heb wat taarten gemaakt";

  const O1 = tokenize(I1, "dutch");
  const O2 = tokenize(I2, "dutch");

  expect(O1).toMatchSnapshot();
  expect(O2).toMatchSnapshot();
});

it("Should stem an array correctly in dutch", async () => {
  // some words in dutch
  const input: string[] = ["banken"];

  // the expected output
  const expected = ["bank"];

  const output = stemArray(input, "dutch");

  expect(output).toStrictEqual(expected);
});

it("Should stem an array correctly in english", async () => {
  // some words in english
  const input: string[] = ["awesome"];

  // the expected output
  const expected = ["awesom"];

  const output = stemArray(input, "english");

  expect(output).toStrictEqual(expected);
});

it("Should stem an array correctly in french", async () => {
  // some words in french
  const input: string[] = ["haussant"];

  // the expected output
  const expected = ["hauss"];

  const output = stemArray(input, "french");

  expect(output).toStrictEqual(expected);
});

it("Should stem an array correctly in italian", async () => {
  // some words in italian
  const input: string[] = ["indicatore"];

  // the expected output
  const expected = ["indic"];

  const output = stemArray(input, "italian");

  expect(output).toStrictEqual(expected);
});

it("Should stem an array correctly in norwegian", async () => {
  // some words in norwegian
  const input: string[] = ["hjemlet"];

  // the expected output
  const expected = ["hjeml"];

  const output = stemArray(input, "norwegian");

  expect(output).toStrictEqual(expected);
});

it("Should stem an array correctly in portugese", async () => {
  // some words in portugese
  const input: string[] = ["velhas"];

  // the expected output
  const expected = ["velh"];

  const output = stemArray(input, "portugese");

  expect(output).toStrictEqual(expected);
});

it("Should stem an array correctly in russian", async () => {
  // some words in russian
  const input: string[] = ["вагоне"];

  // the expected output
  const expected = ["вагон"];

  const output = stemArray(input, "russian");

  expect(output).toStrictEqual(expected);
});

it("Should stem an array correctly in spanish", async () => {
  // some words in spanish
  const input: string[] = ["avenida"];

  // the expected output
  const expected = ["aven"];

  const output = stemArray(input, "spanish");

  expect(output).toStrictEqual(expected);
});

it("Should stem an array correctly in swedish", async () => {
  // some words in swedish
  const input: string[] = ["jemförelser"];

  // the expected output
  const expected = ["jemför"];

  const output = stemArray(input, "swedish");

  expect(output).toStrictEqual(expected);
});
