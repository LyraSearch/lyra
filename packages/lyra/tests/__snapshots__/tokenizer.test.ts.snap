// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should tokenize and stem correctly in english 1`] = `
Set {
  "the",
  "quick",
  "brown",
  "fox",
  "jump",
  "over",
  "lazi",
  "dog",
}
`;

exports[`Should tokenize and stem correctly in english 2`] = `
Set {
  "I",
  "bake",
  "some",
  "cake",
}
`;

exports[`Should tokenize and stem correctly in french 1`] = `
Set {
  "voyon",
  "quel",
  "temp",
  "il",
  "fait",
  "dehor",
}
`;

exports[`Should tokenize and stem correctly in french 2`] = `
Set {
  "j",
  "ai",
  "fait",
  "de",
  "g",
  "teau",
}
`;

exports[`Should tokenize and stem correctly in italian 1`] = `
Set {
  "ho",
  "cucin",
  "dell",
  "tort",
}
`;

exports[`Should tokenize and stem correctly in italian 2`] = `
Set {
  "dorm",
  "una",
  "cos",
  "difficil",
  "quand",
  "i",
  "test",
  "non",
  "pass",
}
`;

exports[`Should tokenize and stem correctly in norwegian 1`] = `
Set {
  "jeg",
  "kokt",
  "noen",
  "kak",
}
`;

exports[`Should tokenize and stem correctly in norwegian 2`] = `
Set {
  "sov",
  "er",
  "en",
  "vansk",
  "ting",
  "n",
  "r",
  "test",
  "mislykk",
}
`;

exports[`Should tokenize and stem correctly in portugese 1`] = `
Set {
  "eu",
  "cozinh",
  "alguns",
  "bol",
}
`;

exports[`Should tokenize and stem correctly in portugese 2`] = `
Set {
  "dorm",
  "uma",
  "cois",
  "dif",
  "cil",
  "quand",
  "os",
  "test",
  "falh",
}
`;

exports[`Should tokenize and stem correctly in russian 1`] = `
Set {
  "я",
  "приготов",
  "пирожн",
}
`;

exports[`Should tokenize and stem correctly in russian 2`] = `
Set {
  "спат",
  "трудн",
  "когд",
  "тест",
  "не",
  "срабатыва",
}
`;

exports[`Should tokenize and stem correctly in swedish 1`] = `
Set {
  "jag",
  "lag",
  "n",
  "gra",
  "kak",
}
`;

exports[`Should tokenize and stem correctly in swedish 2`] = `
Set {
  "att",
  "sov",
  "r",
  "en",
  "sv",
  "sak",
  "n",
  "test",
  "misslyck",
}
`;
